include/gridtools/fn/scan.hpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2021, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+#include <cassert>
+#include <type_traits>
+
+#include "../common/functional.hpp"
+#include "../common/integral_constant.hpp"
+#include "../common/tuple.hpp"
+#include "../sid/concept.hpp"
+
+namespace gridtools::fn {
+    namespace scan_impl_ {
+        template <class F, class Projector = host_device::identity>
+        struct scan_pass {
+            F m_f;
+            Projector m_p;
+            constexpr scan_pass(F f, Projector p = {}) : m_f(f), m_p(p) {}
+        };
+
+        template <class T>
+        struct is_scan_pass : std::false_type {};
+
+        template <class F, class P>
+        struct is_scan_pass<scan_pass<F, P>> : std::true_type {};
+
+        template <bool IsBackward>
+        struct base : std::bool_constant<IsBackward> {
+            static GT_FUNCTION consteval auto prologue() { return tuple(); }
```

Comment:
I removed consteval. Please let me know if you spot another C++20 feature. Currently this test (as all other `fn` stuff) only is enabled if C++20 is enabled. Should I change that already or might it be easier to do that later during development it might be necessary to backport some `fn` stuff to C++17?

Reply:
well, I don't have the strict opinion here.
back porting to c++17 should be strait forward.
My strategic vision was to stick with `c++20`/`clang cuda` as far as possible and back port to `c++17`/`nvcc` only if we will forced.
But I guess that Hannes has the opposite opinion.